wasi/wasi.c

```diff
@@ -1104,8 +1107,33 @@ wasiFDReaddir(
         inode = entry->d_ino;
         name = entry->d_name;
         nameLength = strlen(name);
-        fileType = wasiFileTypeFromDirentFileType(entry->d_type);
+#ifndef __HAIKU__
+        if (entry->d_type != DT_UNKNOWN) {
+            fileType = wasiFileTypeFromDirentFileType(entry->d_type);
+        } else {
+#endif
+            if (lstat(descriptor.path, &entry_stat)) {
+                WASI_TRACE(("fd_readdir: lstat failed: %s", strerror(errno)));
+                return wasiErrno();
+            }
 
+            if (S_ISREG(entry_stat.st_mode)) {
```

Comment:
You can use the existing function `wasiFileTypeFromMode` here

Reply:
Good idea.  I'll make that change right away.